### ISSUE TITLE
Preserve original scope names in exported ONNX graph

### ISSUE THREAD

**Issue by marcocaccin** [pytorch/pytorch#75100]
### 🚀 The feature, motivation and pitch

When a PyTorch model is exported, the graph nodes take uninformative names like `f"{op_type}_{op_cnt}"`, and so do tensors (besides the input and output tensors of the overall graph).
In some contexts, it would be extremely helpful if each node in the exported ONNX graph would be named after the scope name of the PyTorch module from which it came from. Additionally, intermediate tensors should also ideally be named after the node that created them.

Here's a very subjective list of why it matters:

### Visualisation, Debugging

![netron_view](https://user-images.githubusercontent.com/7538983/161244142-65b8e4c5-23be-4daa-8b49-fb12ef1925e7.jpg)

Let's say one goal of exporting to ONNX is to visualise a model (without loss of generality, let's think Netron) to reason about its architecture. For anything but the most trivial NN, this is simply not possible if nodes are named `Conv_{N}` and tensors are `{M}`. 

As soon as the node name becomes `unet.encoder.stage1.conv0` and its output tensor `unet.encoder.stage1.conv0.output0`, I become able to understand what node am I looking at, where is a skip connection going, and this sort of things.

### ONNX graph manipulation

For some people, exporting a model to ONNX is not the end of the story. There may be some graph- or weight- level manipulations that must be applied on sub-graphs defined by the name scopes of their nodes or their intermediate tensors: just to give an idea, a silly example here may be "apply batchnorm fusion only on the decoder of the UNet".

### TensorFlow does it 😛 

Let's be clear: I am by no means a TF fan, but credit is due here. On that side, the feature I'm pitching is the default in [`tf2onnx`](https://github.com/onnx/tensorflow-onnx/tree/main/tf2onnx) and is a design decision that makes a lot of sense.

### Notes

I fully understand that name scopes may lose some of their meaning when running graph-level optimisations during the export if sets of nodes are replaced by new ones and there is no 1:1 relationship. Even in this case, though, it may be possible to assign a "common ancestor name"... or fall back to the current implementation if nothing can be done.

### 1. Fork PyTorch and diverge

Since at least some of the code of the ONNX exporting is in C++, there is no easy monkey patching solution, the library must be recompiled.

### 2. Hack around with a heavy hand

[Someone](https://github.com/pytorch/pytorch/issues/33463#issuecomment-606399944) suggested that it is possible to monkey patch the `_slow_forward` method of `torch.nn.Module` to give meaningful names to the _tensors_ flying around during graph tracing (note: not the nodes, but from output tensor names to node names the step is trivial).
I tried it and it seems to be a bit of a hit and miss, I would need to have a way deeper understanding of the guts of JIT to ensure that 100% of the output tensors actually have a useful name.

Alternatively, or in conjunction, I am also trying to monkey patch `torch.onnx.utils._export()` to use the `inlined_graph`, dump it, and rename the ONNX nodes as a post-processing.

### Additional context

The scope name would be readily available from the `scopeName()` of each `torch._C.Node` of the TorchScript's `inlined_graph`, but the export instead uses the `graph` where such information has been stripped off (for example, [here](https://github.com/pytorch/pytorch/blob/00607e75e5f3157839719efb5b26ca92e91190a3/torch/onnx/utils.py#L430) is the first place where the choice of using the `graph` is made and becomes consequential when exporting a `ScriptModule`).

To make use of this metadata down the line, also the serialisation logic of `torch._C.Graph._export_onnx()` must be adapted. [Here](https://github.com/pytorch/pytorch/blob/00607e75e5f3157839719efb5b26ca92e91190a3/torch/csrc/jit/serialization/export.cpp#L830) is where the nodes of the exported ONNX proto take their names on the fly.

**Comment by alcabilone** [comment]
Hello, is anyone else assigned on this or could i have a try?

**Comment by BowenBao** [comment]
Have you tried exporting with `verbose=True`? It adds stacktrace information to `doc_string` field of each node. https://pytorch.org/docs/stable/onnx.html#torch.onnx.export

**Comment by marcocaccin** [comment]
> Have you tried exporting with `verbose=True`? It adds stacktrace information to `doc_string` field of each node. https://pytorch.org/docs/stable/onnx.html#torch.onnx.export

I tried that first, but TL;DR it simply does not work.
There are a couple of problems with this approach:
1. The docs aren't saying the truth (deserves a separate github issue, but let's not get derailed). To be more specific, the documentation shows that the docstring of each node should contain a `scope:` section (which is indeed what I'm looking for, even if in a slightly different format), but in reality this section does not exist. 
You can easily reproduce this by
```python
    class MyModel(torch.nn.Module):
        def __init__(self):
            super(MyModel, self).__init__()
            self.chunk0 = torch.nn.Sequential(torch.nn.Conv2d(1, 1, 1), torch.nn.BatchNorm2d(1), torch.nn.ReLU())
            self.chunk1 = torch.nn.Sequential(torch.nn.Conv2d(1, 1, 1), torch.nn.ReLU())

        def forward(self, x):
            x = self.chunk0(x)
            x = self.chunk1(x)
            return x

    model_pt = MyModel()
    input_tensor = torch.randn(1, 1, 32, 32)
    torch.onnx.export(
        model_pt,
        input_tensor,
        "model_test.onnx",
        verbose=True,
        input_names=["inputs"],
        output_names=["outputs"],
    )
    model_onnx = onnx.load("model_test.onnx")
    print(onnx.helper.printable_graph(model_onnx.graph))
```
the output will look like
```
graph torch-jit-export (
  %inputs[FLOAT, 1x1x32x32]
) initializers (
  %chunk1.0.weight[FLOAT, 1x1x1x1]
  %chunk1.0.bias[FLOAT, 1]
  %onnx::Conv_16[FLOAT, 1x1x1x1]
  %onnx::Conv_17[FLOAT, 1]
) {
  %input.4 = Conv[dilations = [1, 1], group = 1, kernel_shape = [1, 1], pads = [0, 0, 0, 0], strides = [1, 1]](%inputs, %onnx::Conv_16, %onnx::Conv_17)
  %input.8 = Relu(%input.4)
  %input.12 = Conv[dilations = [1, 1], group = 1, kernel_shape = [1, 1], pads = [0, 0, 0, 0], strides = [1, 1]](%input.8, %chunk1.0.weight, %chunk1.0.bias)
  %outputs = Relu(%input.12)
  return %outputs
}
```

2. Related to (1), one cannot reliably infer the name scopes of every node. Two conditions must be true for that to be possible:
   - The node must have parameters, which is true only for stateful layers such as `Conv` but not for stateless such as `ReLU` or `Add`
   - The node must map 1:1 to a `torch.nn.Module`, otherwise even the names of layer parameters are mangled into something meaningless. As an example, look at `chunk0` from the example above where the Conv+BN layers have been fused.

**Comment by marcocaccin** [comment]
@BowenBao thank you!
A small clarification/nit to be made about your last comment though: the `doc_string` field contains sort of a traceback of where in the code the node comes from, but does not seem to have any reference to the scope name.
I don't have access to the MS link but if there's anything I can help you (e.g., providing a gist of how I solved the problem at the moment) please ping me at any moment 🙂 !

**Comment by garymm** [comment]
Same request from @lutzroeder to make visualizing structure in Netron possible.
Basically, what TensorFlow is doing. Nodes are named "module1/submodule2/nodename_x" and the "/" are interpreted by Netron as hierarchy.

**Comment by BowenBao** [comment]
> Same request from @lutzroeder to make visualizing structure in Netron possible. Basically, what TensorFlow is doing. Nodes are named "module1/submodule2/nodename_x" and the "/" are **interpreted by Netron as hierarchy**.

@garymm, @lutzroeder Is there any example to check how `interpreted by Netron as hierarchy` looks and works? I'm trying to see if similar effect is achieved with updating pytorch node naming.

**Comment by RalphMao** [comment]
This feature will be super helpful!

**Comment by cchan-lm** [comment]
Tried the nightly to test out https://github.com/pytorch/pytorch/pull/82038 and https://github.com/pytorch/pytorch/pull/82040 and it's already very useful! Do you know when these will be in an official PyTorch release?

**Comment by BowenBao** [comment]
@cchan-lm Happy to hear that you find it useful, please feel free to give us feedback! This should be ready in the next PyTorch release 1.14 which is coming soon.

**Comment by marcocaccin** [comment]
This is fantastic @BowenBao, thank you for your hard work and making this feature happen!

**Comment by cchan-lm** [comment]
@BowenBao That is awesome news, and we're looking forward to 1.13! For feedback, should we open another issue or comment here?

**Comment by BowenBao** [comment]
> @BowenBao That is awesome news, and we're looking forward to 1.13! For feedback, should we open another issue or comment here?

Please open new issue, but feel free to mention this one for reference if needed.

**Comment by kevalmorabia97** [comment]
Currently the node names do not correspond to submodule names. For example, if I export the onnx graph for `torchvision.models.resnet18`, I see nodes with names like `/layer1/layer1.0/conv2/Conv` while in the actual model, the corresponding submodule name is `layer1.0.conv2`. Is it possible to have the submodule name in the onnx node name i.e.e `/layer1.0.conv2/Conv`?
